### PR TITLE
Bug 2090182: [Nutanix]Create a machineset with invalid image, machine stuck in "Provisioning" phase

### PR DESCRIPTION
**Pull request by yanhua121**
Fixing Bug 2090182 - [Nutanix]Create a machineset with invalid image, machine stuck in "Provisioning" phase
Added validation of the VM configuration fields in the machine.spec.providerSpec before calling the prism API to create a VM. If the validation failed, an InvalidMachineConfiguration error will return with all the VM configuration errors.

### CODE REVIEW

File: pkg/actuators/machine/vm.go [hunk 1]
```diff
@@ -14,6 +15,146 @@ import (
...
 
+// validateVMConfig verifies if the machine's VM configuration is valid
+func validateVMConfig(mscp *machineScope) field.ErrorList {
+	errList := field.ErrorList{}
+	fldPath := field.NewPath("spec", "providerSpec", "value")
+
+	var err error
+	var errMsg string
+
+	// verify the cluster configuration
```
Comment (thunderboltsid) [inline pkg/actuators/machine/vm.go:26]: Consider Breaking down the validation logic into submethods (i.e. validateClusterConfig , validateImageConfig, validateSubnetConfig)

File: pkg/actuators/machine/vm.go [hunk 2]
```diff
@@ -14,6 +15,146 @@ import (
...
+			errList = append(errList, field.Required(fldPath.Child("cluster", "uuid"), "Missing cluster uuid"))
+		} else {
+			clusterUUID := *mscp.providerSpec.Cluster.UUID
+			if _, err = mscp.nutanixClient.V3.GetCluster(clusterUUID); err != nil {
+				errMsg = fmt.Sprintf("Failed to find cluster with uuid %v. error: %v", clusterUUID, err)
+				errList = append(errList, field.Invalid(fldPath.Child("cluster", "uuid"), clusterUUID, errMsg))
+			}
+		}
+	default:
+		errMsg = fmt.Sprintf("Invalid cluster identifier type, the valid is %q or %q.", machinev1.NutanixIdentifierName, machinev1.NutanixIdentifierUUID)
```
Comment (thunderboltsid) [inline pkg/actuators/machine/vm.go:53]: Substitute "the valid is" with "valid types are"

**Review by elmiko — COMMENTED** [review]
this generally makes sense to me. i agree with @thunderboltsid 's suggestions.

File: pkg/actuators/machine/vm.go (hunk 2 above)
Comment (yanhua121) [reply pkg/actuators/machine/vm.go]: Changed

File: pkg/actuators/machine/vm.go [hunk 3]
```diff
@@ -14,6 +15,146 @@ import (
...
+			errList = append(errList, field.Required(fldPath.Child("image", "uuid"), "Missing image uuid"))
+		} else {
+			imageUUID := *mscp.providerSpec.Image.UUID
+			if _, err = mscp.nutanixClient.V3.GetImage(imageUUID); err != nil {
+				errMsg = fmt.Sprintf("Failed to find image with uuid %v. error: %v", imageUUID, err)
+				errList = append(errList, field.Invalid(fldPath.Child("image", "uuid"), imageUUID, errMsg))
+			}
+		}
+	default:
+		errMsg = fmt.Sprintf("Invalid image identifier type, the valid is %q or %q.", machinev1.NutanixIdentifierName, machinev1.NutanixIdentifierUUID)
```
[the remark this answers is not in the document]
Comment (yanhua121) [reply pkg/actuators/machine/vm.go]: Changed

File: pkg/actuators/machine/vm.go [hunk 4]
```diff
@@ -14,6 +15,146 @@ import (
...
+			if subnet.UUID == nil || *subnet.UUID == "" {
+				errList = append(errList, field.Required(fldPath.Child("subnet").Child("uuid"), "Missing subnet uuid"))
+			} else {
+				if _, err = mscp.nutanixClient.V3.GetSubnet(*subnet.UUID); err != nil {
+					errMsg = fmt.Sprintf("Failed to find subnet with uuid %v. error: %v", *subnet.UUID, err)
+					errList = append(errList, field.Invalid(fldPath.Child("subnet", "uuid"), *subnet.UUID, errMsg))
+				}
+			}
+		default:
+			errMsg = fmt.Sprintf("Invalid subnet identifier type, the valid is %q or %q.", machinev1.NutanixIdentifierName, machinev1.NutanixIdentifierUUID)
```
[the remark this answers is not in the document]
Comment (yanhua121) [reply pkg/actuators/machine/vm.go]: Changed

**Review by elmiko — COMMENTED** [review]
thanks for the updates @yanhua121 

/approve

File: pkg/actuators/machine/actuator_test.go [hunk 5]
```diff
@@ -316,7 +302,9 @@ func TestMachineEvents(t *testing.T) {
 
 			gs.Eventually(waitForEvent, timeout).Should(Succeed())
 
-			gs.Expect(eventList.Items[0].Message).To(ContainSubstring(tc.event))
+			for _, msg := range tc.events {
+				gs.Expect(eventList.Items[0].Message).To(ContainSubstring(msg))
```
[the remark this answers is not in the document]
Comment (yanhua121) [reply pkg/actuators/machine/actuator_test.go:306]: In each test case, only the top event is supposed to check whether containing the specified messages. So here we don't need to iterate the eventList items.

[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/actuators/machine/actuator_test.go:306]: How many events are we expecting on each test? Perhaps if we only expect a single event to be produced we can assert that above, else I'd like to know what those extra events are

[the remark this answers is not in the document]
Comment (yanhua121) [reply pkg/actuators/machine/actuator_test.go:306]: See lines 293-295 as below, it verifies there should only be one event item:
    if len(eventList.Items) != 1 {
	return fmt.Errorf("expected len 1, got %d", len(eventList.Items))
    }

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

Thanks!

**Pull request merged**